driver-core/src/main/com/mongodb/connection/AsynchronousSocketChannelStreamFactoryFactory.java

```diff
@@ -16,15 +16,89 @@
 
 package com.mongodb.connection;
 
+import com.mongodb.internal.connection.PowerOfTwoBufferPool;
+
+import java.nio.channels.AsynchronousChannelGroup;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
 /**
  * A {@code StreamFactoryFactory} implementation for AsynchronousSocketChannel-based streams.
  *
  * @see java.nio.channels.AsynchronousSocketChannel
  * @since 3.1
  */
 public class AsynchronousSocketChannelStreamFactoryFactory implements StreamFactoryFactory {
+    private final BufferProvider bufferProvider;
+    private final AsynchronousChannelGroup group;
+
+    /**
+     * Construct an instance with the default {@code BufferProvider} and {@code AsynchronousChannelGroup}.
+     *
+     * @deprecated Use {@link AsynchronousSocketChannelStreamFactoryFactory#builder()} instead to construct the
+     * {@code AsynchronousSocketChannelStreamFactoryFactory}.
+     */
+    @Deprecated
+    public AsynchronousSocketChannelStreamFactoryFactory() {
+        this(builder());
+    }
+
+    /**
+     * Gets a builder for an instance of {@code AsynchronousSocketChannelStreamFactoryFactory}.
+     * @return the builder
+     * @since 3.6
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * A builder for an instance of {@code AsynchronousSocketChannelStreamFactoryFactory}.
+     *
+     * @since 3.6
+     */
+    public static final class Builder {
+        private BufferProvider bufferProvider;
+        private AsynchronousChannelGroup group;
+
+        /**
+         * Sets the bufferProvider
+         *
+         * @param bufferProvider the bufferProvider
+         * @return this
+         */
+        public Builder bufferProvider(final BufferProvider bufferProvider) {
```

Comment:
I would prefer to leave the BufferProvider as unconfigurable until some asks for it.  The ByteBuf interface on which BufferProvider relies is not wonderful, and exposing it via the public API will make it harder to change it without affecting more applications than we would if we don't make it configurable in the high-level API.

Reply:
No problems - removed it from the config.